### PR TITLE
Add crypto_period_seconds to Widevine key request

### LINKED ISSUE #544: Missing crypto_period_seconds information in Widevine key request with key rotation
Hi,

When using key rotation, the packager does not send crypto_period_seconds information in the key request for Widevine DRM. The request only has crypto_period_index and _count. However, the crypto period is necessary for a key server to relate generated keys to the media playback time. This helps server to provide relevant keys to a client during license request. 

I have prepared a patch to fix this issue. I will make a pull request soon.

### PR DESCRIPTION
**Pull request by srebrnyp**
When using key rotation with Widevine DRM, a key server has to know
the duration of crypto period to relate generated keys to the media
playback time. This helps the server to provide relevant keys to
a client during license request.

Closes #544.

### CODE REVIEW

File: packager/media/base/key_source.h [hunk 1]
```diff
@@ -77,11 +77,14 @@ class KeySource {
   /// Get encryption key of the specified track type at the specified index.
   /// @param crypto_period_index is the sequence number of the key rotation
   ///        period for which the key is being retrieved.
+  /// @param crypto_period_duration_in_seconds is the duration of the crypto
+  ///        period in seconds
```
Comment (kqyang) [inline packager/media/base/key_source.h]: nit: Add an ending dot ".".

File: packager/media/base/widevine_key_source_unittest.cc [hunk 2]
```diff
@@ -514,7 +514,7 @@ const char kCryptoPeriodRequestMessageFormat[] =
     R"("tracks":[{"type":"SD"},{"type":"HD"},{"type":"UHD1"},)"
     R"({"type":"UHD2"},{"type":"AUDIO"}],)"
     R"("drm_types":["WIDEVINE"],)"
-    R"("first_crypto_period_index":%u,"crypto_period_count":%u,)"
+    R"("first_crypto_period_index":%u,"crypto_period_count":%u,"crypto_period_seconds":%u,)"
```
Comment (kqyang) [inline packager/media/base/widevine_key_source_unittest.cc]: Can you break it to two lines so the line length is less than 80 characters per line?

i.e.

R"("first_crypto_period_index":%u,"crypto_period_count":%u,)"
R"("crypto_period_seconds":%u,)"

File: packager/media/base/widevine_key_source.cc [hunk 3]
```diff
@@ -213,13 +213,17 @@ Status WidevineKeySource::GetKey(const std::vector<uint8_t>& key_id,
...
                                              const std::string& stream_label,
                                              EncryptionKey* key) {
   DCHECK(key_production_thread_.HasBeenStarted());
   // TODO(kqyang): This is not elegant. Consider refactoring later.
   {
     base::AutoLock scoped_lock(lock_);
     if (!key_production_started_) {
+      // Record crypto period duration, which should not change once
+      // the key rotation has been enabled.
+      crypto_period_duration_in_seconds_ = crypto_period_duration_in_seconds;
```
Comment (kqyang) [inline packager/media/base/widevine_key_source.cc]: Can you return an error if |crypto_period_duration_in_seconds_| changes, i.e.

    if (!key_production_started_) {
      ...
      crypto_period_duration_in_seconds_ = crypto_period_duration_in_seconds;
      ...
    } else if (crypto_period_duration_in_seconds_ !=
               crypto_period_duration_in_seconds) {
      return Status(error::INVALID_ARGUMENT,
                    "Crypto period duration should not change.");
    }

Comment (srebrnyp) [reply packager/media/base/widevine_key_source.cc]: Yes, this is much better than just a comment

**Comment by kqyang** [comment]
@srebrnyp The changes have been merged. Thanks again for your help in improving Shaka Packager!

**Pull request merged**